### ISSUE TITLE
Randomize TIA on startup (was: Issue with "latched input ports" (VBLANK bit 6))

### ISSUE THREAD

**Issue by ale-79** [stella-emu/stella#126]
On the first write to VBLANK, if bit 6 is set (to enable "latched input ports"), then both INPT4 and INPT5 have bit 7 cleared even if the firebuttons on the joysticks were not pressed. **This only happens on the very first write to VBLANK**; all subsequent changes to bit 6 behave as expected. (it took me a while to understand why I couldn't reproduce the issue in a test rom, because I was using the standard CLEAN_START macro to initialize ram and TIA at the beginning of my code...)

This happens also in Stellerator, but not on Stella 4.x.

I guess this is harmless for almost every game, but at least the hack of "Yahtzee" in the last post of this thread on AA is affected:
[http://atariage.com/forums/topic/4374-yahtzee-for-2600/page-2#entry3338324](http://atariage.com/forums/topic/4374-yahtzee-for-2600/page-2#entry3338324)

The game initializes all the TIA registers except RSYNC,WSYNC,VBLANK and VSYNC, and the first time it writes to VBLANK it sets the "latch" bit (I don't remember why I added that,  and I can't find the source...). 
If you start Stella in debugger mode and set a breakpoint at $11CE, you'll see that bit 7 of INPT4 and INPT5 are set (no button pressed), while they both get cleared as soon as VBLANK is written to. The result is that the game skips the title screen as if the joystick button was pressed. This only happens when you start the rom. The title screen is displayed if you press RESET.

I tried loading the rom in single-game mode on the harmony cart and tested it on a 2600 Jr, so that there's no 7800 bios or harmony menu executed before the game. The result is that _sometimes_ it skips the title screen. I guess that if the latch state is random at power on, but that's still different from the current Stella behaviour. 

I believe fixing this should be VERY low priority. I just thought it was worth reporting it.

**Comment by sa666666** [comment]
So to be clear, this worked in Stella 4.x and doesn't now?  Because if so, it's a regression that probably should be fixed.

**Comment by ale-79** [comment]
Yes. The rom itself should be fixed to correctly initialize VBLANK, but the fact that writing to VBLANK changes INPT4 and INPT5 without pressing the joystick buttons is a bug which wasn't there in stella 4.x.

The fact that it only happens on the first write to VBLANK makes it extremely unlikely that it would affect a well programmed game. That's why I think it should be considered low priority.

**Comment by DirtyHairy** [comment]
I'll look into it later this week.

**Comment by ale-79** [comment]
> [...] but the fact that writing to VBLANK changes INPT4 and INPT5 without pressing the joystick buttons is a bug which wasn't there in stella 4.x.

After sleeping on it, I think this is not a bug after all, but the result of TIA registers not being randomized in the emulator (and the state of some of them is different in Stella 5 compared to previous versions).

According to the Stella Programmer's Guide, setting VBLANK D6=0 disable latches (reading INPTx returns the actual input pin state) and reset them to logic 1, 
while D6=1 enable them (reading INPUTx returns the latches' value) but _doesn't change their state_.

By looking at the debugger, in Stella all registers are cleared at power on, and in particular:

- both latches are at logic 0
- they're disabled (so INPTx returns the actual joystick button state)

Then on the first write to VBLANK 

if D6=0 -> both latches are reset to 1 and they're still disabled
if D6=1 -> latches are enabled and since they're set to 0, that's what a read from INPUTx will return. 

So the emulation is correct, altough on real hardware the initial state is random, so this is just one of the possible scenarios.

I guess we can close this one (the 'bug' is in the game as it relies on an unitializated value and it sometimes glitches on real hardware too), 

Maybe an option to randomize all the TIA registers (like is already in place for ram and cpu) enabled by default, would more closely represent real hardware.

**Comment by sa666666** [comment]
Making the TIA registers random at startup seems easy enough, but I will wait on @DirtyHairy feedback before I proceed.

**Comment by DirtyHairy** [comment]
Fully randomization of the initial state is tricky. However, I think what would be most useful is randomizing those quantities that can be read back (essentially collisions and latched / dumped input ports), and that should be pretty straightforward. I'll take a look at it, but I don't think we should make this a prerequesite for 5.0 --- it's one of those things that might start out easy and turn out to be a bottomless pit :smirk:

**Comment by sa666666** [comment]
I agree on not making it a priority for 5.0, since the release is stretching on at this point.  The original intent was for 5.0 to be basically 4.x + the new TIA core, but it has expanded quite a bit beyond that.  Not a bad thing, really, since the extra features we've added are very cool.  But we need to do the 5.0 release at some point, or it will stretch to the end of the 2017.  I plan to do 5.0 but the end of this month.  And I haven't even gotten to the PAL colour-loss or phosphor stuff yet ...

**Comment by DirtyHairy** [comment]
@sa666666 perhaps we should start labelling issues as 5.0 or post-5.0 :label:

**Comment by sa666666** [comment]
I already have in a few issues, simply by adding a message stating as such.  Is there a more formal way of doing this in github?

**Comment by DirtyHairy** [comment]
The thing that you can do is to create a tag that labels all issues that are targeted for a specific release. I have done so for the two bugs that you have mentioned. Good that you remind me about jitter; I had forgotten about it. I'll look into it the next time when I have spare time (this weekend I guess).

**Comment by thrust26** [comment]
@DirtyHairy Can this be closed?

**Comment by DirtyHairy** [comment]
Nope, let's keep it open --- initial state randomization is a  valid issue imho :smirk:

**Comment by sa666666** [comment]
Should this be renamed to something else (indicating it's more about randomization now)?  And should we remove it from specific release tags, and leave it as a general issue?

**Comment by thrust26** [comment]
>Full randomization of the initial state is tricky as it does not map back to a set of register values (it's mostly a plethora of counters and a bunch of flags, and not all naive combinations are valid). However, I think what would be most useful is randomizing those quantities that can be read back (essentially collisions and latched / dumped input ports), and that should be pretty straightforward. I'll take a look at it, but I don't think we should make this a prerequesite for 5.0 --- it's one of those things that might start out easy and turn out to be a bottomless pit 😏

@DirtyHairy I suppose the `reset` methods of the various display objects and inputs would have to use random values where appropriate? 

Maybe I have an easier idea. How about poking a few 1000 random values in random order to the TIA addresses ($00..$2F)? Shouldn't that create sufficient randomness too?

**Comment by DirtyHairy** [comment]
Random pokes are a good idea. The issue with directly randomizing during `reset()` are implicit constraints between the randomized values --- not all combinations are valid.

**Comment by thrust26** [comment]
> The issue with directly randomizing during reset() are implicit constraints between the randomized values --- not all combinations are valid.

That's exactly why I came up with my idea. Here the existing code checks the constrains.

**Comment by DirtyHairy** [comment]
We just need to be careful with 3E/3F .

**Comment by thrust26** [comment]
Maybe, we have to check the initialization order and the `Cart` classes.

**Comment by thrust26** [comment]
Should we define an option like for RAM randomization? Or should we always do it? Maybe some old games might break from this.

**Comment by thrust26** [comment]
The cart is reset after the TIA, so we should not expect any problems here

Also I checked all TigerVison $3F games and our $3E(+) test ROMs. If they work without randomization, they work with too.

**Comment by thrust26** [comment]
@DirtyHairy Shall we close this one now, or do want to check/improve my randomization yourself?

**Comment by thrust26** [comment]
Finalized with 79eb8656

**Comment by ale-79** [comment]
Great to have TIA randomization working!

BTW, I think #249 is a duplicate of this one.

**Comment by thrust26** [comment]
> Great to have TIA randomization working!

Thanks. It turned out to be much easier than we thought it would be. 

> BTW, I think #249 is a duplicate of this one.

Yes and no. @DirtyHairy mentioned the RIOT.
